pkg/cache/client/round_tripper.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"net/http"
+	"regexp"
+	"strings"
+
+	"k8s.io/client-go/rest"
+
+	clientshard "github.com/kcp-dev/kcp/pkg/cache/client/shard"
+)
+
+var (
+	// matches shards/name/remainder, capturing name.
+	//
+	// Example: shards/name/remainder
+	// Example: /shards/name/remainder
+	// Example: prefix/shards/name/remainder
+	shardNameRegex = regexp.MustCompile(`shards/([^/]+)/.+`)
+)
+
+// WithShardRoundTripper wraps an existing config's with ShardRoundTripper.
+//
+// Note: it is the caller responsibility to make a copy of the rest config
+func WithShardRoundTripper(cfg *rest.Config) *rest.Config {
+	cfg.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+		return NewShardRoundTripper(rt)
+	})
+
+	return cfg
+}
+
+// ShardRoundTripper is a shard aware wrapper around http.RoundTripper.
+// It changes the URL path to target a shard from the context.
+//
+// For example given "amber" shard name in the context it will change
+// apis/apis.kcp.dev/v1alpha1/apiexports to /shards/amber/apis/apis.kcp.dev/v1alpha1/apiexports
+type ShardRoundTripper struct {
+	delegate http.RoundTripper
+}
+
+// NewShardRoundTripper creates a new shard aware round tripper
+func NewShardRoundTripper(delegate http.RoundTripper) *ShardRoundTripper {
+	return &ShardRoundTripper{
+		delegate: delegate,
+	}
+}
+
+func (c *ShardRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	shard := ShardFromContext(req.Context())
+	if !shard.Empty() {
+		req = req.Clone(req.Context())
+		path, err := generatePath(req.URL.Path, shard)
+		if err != nil {
+			return nil, err
+		}
+		req.URL.Path = path
+
+		rawPath, err := generatePath(req.URL.RawPath, shard)
+		if err != nil {
+			return nil, err
+		}
+		req.URL.RawPath = rawPath
+	}
+	return c.delegate.RoundTrip(req)
+}
+
+// generatePath formats the request path to target the specified shard
+func generatePath(originalPath string, shard clientshard.Name) (string, error) {
+	// if the originalPath already has the shard then the path was already modified and no change needed
+	if strings.Contains(originalPath, shard.Path()) {
+		return originalPath, nil
+	}
+	// if the originalPath already has a shard set just overwrite it to the given one
+	if strings.Contains(originalPath, "/shards") {
```

Comment:
Is this specifically not doing a prefix check for some reason?

Reply:
as of today, we could actually check for a prefix, maybe in the future, it will have to be changed.

Reply:
done